### PR TITLE
Delimited data Parser in specification with App Builder Roadmap.

### PR DESCRIPTION
**Pull request by pm48**

### CODE REVIEW

File: demos/dimensions/src/main/java/com/datatorrent/demos/dimensions/ads/ParserApplication.java [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+    parser.setFieldDelimiter(',');
+    parser.setLineDelimiter("\n");
+    @SuppressWarnings("unchecked")
+    DevNull<Map<String,Object>> devNull = dag.addOperator("DevNull", DevNull.class);
+
+    dag.addStream("Kafka2Parser", kafkaStringInput.outputPort, parser.input);
+    dag.addStream("Parser2DevNull",parser.output,devNull.data);
+  }
+
+  public static class KafkaSinglePortStringInputOperator extends AbstractKafkaSinglePortInputOperator<byte[]>
```
Comment (tweise) [inline demos/dimensions/src/main/java/com/datatorrent/demos/dimensions/ads/ParserApplication.java]: This operator should be named appropriately and be in the contrib package.

File: library/src/main/java/com/datatorrent/lib/parser/AbstractParser.java [hunk 2]
```diff
@@ -0,0 +1,298 @@
...
+  @NotNull
+  protected int fieldDelimiter;
+  protected String lineDelimiter;
+  protected transient String[] properties;
+  protected transient CellProcessor[] processors;
+  protected boolean isHeader;
+
+  public enum FIELD_TYPE
+  {
+    BOOLEAN, DOUBLE, INTEGER, FLOAT, LONG, SHORT, CHARACTER, STRING, DATE, UNKNOWN
```
Comment (tweise) [inline library/src/main/java/com/datatorrent/lib/parser/AbstractParser.java]: Why is the a type "UNKNOWN" - should we not assume String when nothing was specified?

File: library/src/main/java/com/datatorrent/lib/parser/Parser.java [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+import org.supercsv.io.CsvMapReader;
+import org.supercsv.io.ICsvMapReader;
+import org.supercsv.prefs.CsvPreference;
+
+/*
+ * An implementation of AbstractParser which emits a map.
+ * The output is a map with key being the field name supplied by user and value being the value of that field
+ * from input stream.
+ */
+public class Parser extends AbstractParser<Map<String, Object>>
```
Comment (tweise) [inline library/src/main/java/com/datatorrent/lib/parser/Parser.java]: It should be CsvParser, "Parser" is too generic.

**Comment by pm48** [comment]
Renamed CsvParser files and moved them to contrib.Incorporated changes from review in same branch.

**Comment by gauravgopi123** [comment]
I remember Siyuan did some work on CSVParser to use custom reusable reader. Please get it reviewed by him

File: contrib/src/main/java/com/datatorrent/contrib/parser/AbstractCsvParser.java [hunk 4]
```diff
@@ -0,0 +1,295 @@
...
+  {
+    @Override
+    public void process(byte[] tuple)
+    {
+      ICsvReader csvReader = null;
+      CsvPreference preference = new CsvPreference.Builder('"', fieldDelimiter, lineDelimiter).build();
+      InputStream in;
+      BufferedReader br = null;
+      try {
+        in = new ByteArrayInputStream(tuple);
```
Comment (siyuanh) [inline contrib/src/main/java/com/datatorrent/contrib/parser/AbstractCsvParser.java]: We can copy code(ReusableStringReader) from https://github.com/DataTorrent/Olive/blob/fd431c83878ebd94d13ecf08465521af97a221e2/pocs/poc1/src/main/java/com/datatorrent/lib/parser/CsvParserOperator.java 
to avoid re-create reader object for each tuple

**Comment by pm48** [comment]
Added the option for user to enter mapping of field names and data type through a hdfs file.
Also used a custom reusable string reader.

File: contrib/src/main/java/com/datatorrent/contrib/parser/AbstractCsvParser.java [hunk 5]
```diff
@@ -0,0 +1,430 @@
...
+
+  /**
+   * This input port receives byte array as tuple.
+   */
+  public final transient DefaultInputPort<byte[]> input = new DefaultInputPort<byte[]>()
+  {
+    @Override
+    public void process(byte[] tuple)
+    {
+      //InputStream in;
```
Comment (tweise) [inline contrib/src/main/java/com/datatorrent/contrib/parser/AbstractCsvParser.java]: Why is this here?

File: contrib/src/main/java/com/datatorrent/contrib/parser/AbstractCsvParser.java [hunk 6]
```diff
@@ -0,0 +1,430 @@
...
+  }
+
+  @Override
+  public void teardown()
+  {
+    try {
+      csvReader.close();
+    }
+    catch (IOException e) {
+      logger.error("Parsing csv error", e);
```
Comment (tweise) [inline contrib/src/main/java/com/datatorrent/contrib/parser/AbstractCsvParser.java]: Don't swallow exception.

File: contrib/src/main/java/com/datatorrent/contrib/parser/CsvParser.java [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+import org.supercsv.io.CsvMapReader;
+import org.supercsv.io.ICsvMapReader;
+import org.supercsv.prefs.CsvPreference;
+
+/*
+ * An implementation of AbstractParser which emits a map.
+ * The output is a map with key being the field name supplied by user and value being the value of that field
+ * from input stream.
+ */
+public class CsvParser extends AbstractCsvParser<Map<String, Object>>
```
Comment (tweise) [inline contrib/src/main/java/com/datatorrent/contrib/parser/CsvParser.java]: Perhaps CsvToMapParser would clearly indicate the restriction? Let's move this to the demo package for now.

File: demos/dimensions/src/main/resources/META-INF/properties.xml [hunk 8]
```diff
@@ -92,4 +92,9 @@
     <value>#/ops/apps/{appId}/appdata</value>
   </property>
 
+<!--FieldMappingFile for CsvParserApplication-->
```
Comment (tweise) [inline demos/dimensions/src/main/resources/META-INF/properties.xml]: What is this doing here?

File: demos/dimensions/src/main/java/com/datatorrent/demos/dimensions/ads/CsvParserFileMappingInputApplication.java [hunk 9]
```diff
@@ -0,0 +1,148 @@
...
+      fsOutStream.close();
+    }
+    catch (IOException ex) {
+      DTThrowable.rethrow(ex);
+
+    }
+    logger.debug("Written data to HDFS file.");
+  }
+
+  public static class KafkaSinglePortStringInputOperator extends AbstractKafkaSinglePortInputOperator<byte[]>
```
Comment (tweise) [inline demos/dimensions/src/main/java/com/datatorrent/demos/dimensions/ads/CsvParserFileMappingInputApplication.java]: name does not reflect restriction

File: demos/dimensions/src/main/resources/META-INF/properties.xml (hunk 8 above)
Comment (pm48) [inline demos/dimensions/src/main/resources/META-INF/properties.xml]: This is the fieldmappingFile given by user as a property.
Value can be empty in property files since its not null.

File: contrib/src/main/java/com/datatorrent/contrib/parser/CsvParser.java (hunk 7 above)
Comment (pm48) [inline contrib/src/main/java/com/datatorrent/contrib/parser/CsvParser.java]: I created the two parser applications in demos/dimensions/ads. Should CsvToMapParser be moved to same package in demos?
